base/deprecated.jl

```diff
@@ -56,41 +56,33 @@ macro deprecate(old,new)
     end
 end
 
+const have_dep_warned = Set()
+
 function depwarn(msg, funcsym)
     opts = JLOptions()
-    if opts.depwarn > 0
-        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+    if opts.depwarn == 1  # raise a warning
         fn = String(unsafe_load(cglobal(:jl_filename, Ptr{Cchar})))
+        ln = Int(unsafe_load(cglobal(:jl_lineno, Cint)))
+        key = hash(string(fn, ln, msg))
+        (key in have_dep_warned) && return
+        push!(have_dep_warned, key)
         bt = backtrace()
         caller = firstcaller(bt, funcsym)
-        if opts.depwarn == 1 # raise a warning
-            warn(msg, once=(caller != C_NULL), key=caller, bt=bt,
-                 filename=fn, lineno=ln)
-        elseif opts.depwarn == 2 # raise an error
-            throw(ErrorException(msg))
-        end
+        warn(msg, once=(caller != C_NULL), key=caller, bt=bt, filename=fn, lineno=ln)
+    elseif opts.depwarn == 2  # raise an error
+        throw(ErrorException(msg))
     end
 end
 
 function firstcaller(bt::Array{Ptr{Void},1}, funcsym::Symbol)
     # Identify the calling line
-    i = 1
-    while i <= length(bt)
+    for i in 1:length(bt)
         lkups = StackTraces.lookup(bt[i])
-        i += 1
         for lkup in lkups
-            if lkup === StackTraces.UNKNOWN
-                continue
-            end
-            if lkup.func == funcsym
-                @goto found
-            end
+            lkup === StackTraces.UNKNOWN && continue
+            lkup.func == funcsym && return bt[i]
```

Comment:
Isn't this broken ? It used to return the frame containing the call to the deprecated function and it looks like it's returning the deprecated frame itself.